TIME/TIMESTAMP W/O TIME ZONE semantics fix (#7122) - continuation

fiedukow opened the pull request:
This is continuation of #7378 and more complete solution for #7122.
Please follow this doc for detailed documentation of changes applied:
https://docs.google.com/document/d/1UUDktZDx8fGwHZV4VyaEDQURorFbbg6ioeZ5KMHwoCk/edit?usp=sharing
Commenting is enabled one this doc, so please feel free to express your concerns.

CC: @martint @haozhun @dain @losipiuk @findepi @ilfrin

fiedukow commented:
Rebased to current master.

findepi commented:
LGTM. Good job, @fiedukow! And thanks for keeping up with me in #7378 :)
Now to @haozhun or @martint .

fiedukow commented:
Rebased to master & fixed conflits.

fiedukow commented:
Following action where taken:
 * Updated PR with notes from doc describing changes:
    - change in TIMESTAMP -> TIMESTAMP WITH TIME ZONE cast semantic
    - restored to_iso8601 with new semantics
 * Applied changes to the doc itself.
 * Rebased PR to current master (resolved conflicts)

Are we at the point to start review or do you need something more @haozhun ?

fiedukow commented:
Rebased to master. Resolved yet another set of conflicts.

haozhun commented:
I'll start review of this pull request

haozhun commented:
The commits in this pull request is not well organized. For example:

* First commit `Fix current_time, localtime & localtimestamp semantics` referenced `valueToSessionTimeZoneOffsetDiff`, which is not in the first commit
* Test for first commit is in second commit "Improve current_date function testing"

haozhun commented:
@dain told me that it was a deliberate decision that TIME W TIMEZONE are represented as local millis + TZ instead of UTC millis + TZ. This is so that we are can have an accurate representation of TIME W TIMEZONE independent of date even if the zone is a political one. Can you briefly explain why you are changing this?

In addition, we are interested in understanding more why you are particularly interested in changing the behavior of Time With Timezone from one broken state to another (less) broken state. Your fix for TIME W TIMEZONE is much wider than expected. We would like to see changes for TIMESTAMP (Without Timezone) and TIME W TIMEZONE separate.

It would be really nice if you can make sure that your commits are cleanly organized. Right now, they seem to mix up pretty badly.

fiedukow commented:
Sorry for the delay.

I reorganised PR completely:
 - Rebase to current master was performed, conflicts were resolved.
 - Lots of commits were squashed into related groups. Namely tests are in commits that contain changing code now.
 - Order of commits is changed so it is more logical - first we introduce the flag and then develop things behind the flag finishing with changing default of flag to new value.
 - Some leftover commits are left as were after “Split TestDateTimeOperators into variants based on isLegacyTimestamp”, I assumed that those are not worth backward engineering to put them in the exactly right place. So please don’t be mad at me for leaving that as is :) I also do realise that having this split at the beginning of PR would make things much easier, not sure if it is worth reorganising again at this point.
 - I applied both of yours comments above (you were right about both of those things).
 - I double checked if each commits both compiles and also if it passes all Travis tests.
 - I’ve added first commit fixing current master (legacy) current_time and local time only. It is still broken after this patch, but only in a expected way of current logic - time zone offset is taken from January first of 1970, so it can return wrong results on DST timezones or time zones with changes in policies. Before this commit it wasn’t even returning valid time internally (UTC time wasn’t really UTC). Also change in tests may look weird, but the truth of the matter is that those tests ale simply poorly designed. I think the tests I’ve came up with in "Fix current_time, localtime & localtimestamp semantics”, are much more reliable and don’t depend on copying implementation of tested method.



Now to the changes in TIME WITH TIME ZONE, I thought it was discussed in docs already and we’ve agreed on this approach.
My main argument is to keep it as close to standard as possible without vast changes that would break SPI compatibility. Also important bit is that what we currently have is very unintuitive for users.
As for the argument that UTC mills + TZ is better representation I really have doubts about that. Note that TIME (w TZ or not) is not really representation of any particular point in time. It’s really only what we read from clock plus INDEPENDENT time zone information. I don’t see any reason why we should have magical reference point (UTC) losing real type information.

Please note that you can use TIME WITH TIME ZONE in basically two contexts:
 1. With DATE obtained from other source - in this scenario, you can have real UTC value in this date without any problems, so it seems like new approach is fine. For example think about having TIMESTAMP WITH TIME ZONE, than splitting it into DATE and TIME WITH TIME ZONE and then combining back. You would rather expect the value at the end to be same as at the beginning? Right? 
 2. Without any DATE at all - than you may argue that you can still compare events based on UTC time they occurred in. The problem is that this UTC time value in current approach is simply wrong. The value is not used in a context of a date at this point, but was comping from some date when it was inserted, and offset to UTC should come from this particular date, to allow reasonable comparison between values. It doesn’t matter if you use TIMESTAMP as source of this values or if you simply type literals manually. I think is safe to assume that if you type: TIME ’10:00:00 Asia/Kathmandu’ you mean “UTC” value NOW and not 50 years ago. For example think about query: “SELECT TIME ’10:00:00 Asia/Kathmandu’ > TIME ’14:30:00 Europe/Warsaw’;” Do you think anyone would prefer to get which hour was first in 1970 January (never useful IMHO) or now (at least possibly useful)?

This is why I think it’s worth to change TIME WITH TIME ZONE type to much less broken state (basically current version is not usable).
More about details in docs.

Back to you @haozhun

cc: @losipiuk, @findepi

haozhun commented:
I don't think I followed everything in your post.

It looks like you are arguing against using "UTC millis + TZ" to represent "TIME w TZ". But that's what you did as far as I can tell.  If I'm understanding correctly, you changed implementation of "TIME w TZ" from "local millis + TZ" to "UTC millis + TZ". You also similarly changed "TIME w/o TZ" from "local millis" to "UTC millis".

haozhun commented:
@fiedukow 

I understand that:

* We need to make sure `TIME '10:00:00 Asia/Kathmandu'` has a valid representation. This is useful for separating TIMESTAMP w TZ into DATE and TIME w TZ. It is also useful in merging DATE and TIME w TZ into TIMESTAMP w TZ.

I have a few questions:
* Are you ok with making the following expressions fail?
  * `SELECT TIME '10:00:00 Asia/Kathmandu' > TIME ’14:30:00 Europe/Warsaw’`
  * `SELECT TIME '03:00:00 +04:00' > TIME '23:00:00 -04:00'`
  * `SELECT TIME '03:00:00 +04:00' - TIME '23:00:00 -04:00'`
  * (Basically, fail any comparison and math operation on TIME w TZ if TZ does not match.)
* What is your intended internal representation for?
  * `02:30 America/Los_Angeles` and `03:30 America/Los_Angeles` in all 4 cases: ordinary non-DST date, ordinary DST date, both DST cutoff dates. (2am-3am is the duplicate/missing time range)

fiedukow commented:
Ok, I think I was unclear about this TIME W TZ representation. I'll try again :)
# One more explination
Both current master and new version does represent TIME W TZ as pair of `unified miliseconds` and `time zone id`. Now the difference is in what the "unified" means, because in both cases it is misleadingly called UTC.
BEFORE: Unified time is number of milliseconds since midnight MINUS time zone offset calculated from January 1 1970.
AFTER: Unified time is number of milliseconds since midnight MINUS time zone offset lazy calculated when required using information available in context when it's required (namely session start time).

The consequences are as follows:
 - Before, we were able to precisely convert TIME W TZ to other time zone without adding any other information and interpretation holds.
 - After, we have same behaviour for TIME W TZ and TIMESTAMP W TZ type, allowing conversion between them without losing informations.
 - After, we are able to do valid conversion providing date information from outside.
 - After, we are able to always return valid time from current_time, without breaking the semantic of the type.
 - So basically we change precise “UTC” time, to precise "local”, which means that semantically we represent `local wall time + TZ`, before we represented `UTC wall time (with 1970 offset applied) + TZ`.

# Questions answered
 1. Yes this value must be allowed and it should be possible to split and merge data types.
   - eg. `with TS (t) AS (select cast(timestamp '2017-06-07 10:00:00.000 Europe/Warsaw' as time with time zone) t) select from_iso8601_timestamp('2017-06-07T' || cast(hour(t) as varchar) || ':' || cast(minute(t) as varchar) || ':' || cast(second(t) as varchar)) from TS;`
   - BEFORE: `2017-06-07 09:00:00.000 Europe/Warsaw`
   - AFTER: `2017-06-07 10:00:00.000 Europe/Warsaw`
   - This is caused by cast TIMESTAMP WITH TIME ZONE -> TIME WITH TIME ZONE.
 2. Here are precise answers you've requested (Session in America/Los_Angeles TZ), I've added commit with tests showing that this is actually the case:

- BEFORE TEST REPRESENTATION TABLE

| Literal                            | Session in DST date   | Session in non-DST date   | session in non-DST -> DST date | session in DST -> non-DST |
| ---------------------------------- | --------------------- | ------------------------- | ------------------------------ | ------------------------- |
| `TIME '02:30'`                     | `37800000`            | `37800000`                | `37800000`                     | `37800000`                |
| `TIME '03:30'`                     | `41400000`            | `41400000`                | `41400000`                     | `41400000`                |
| `TIME '02:30 America/Los_Angeles'` | `37800000`            | `37800000`                | `37800000`                     | `37800000`                |
| `TIME '03:30 America/Los_Angeles'` | `41400000`            | `41400000`                | `41400000`                     | `41400000`                |

- AFTER CHANGE REPRESENTATION TABLE

| Literal                            | Session in DST date   | Session in non-DST date   | session in non-DST -> DST date | session in DST -> non-DST |
| ---------------------------------- | --------------------- | ------------------------- | ------------------------------ | ------------------------- |
| `TIME '02:30'`                     | `9000000`             | `9000000`                 | `9000000`                      | `9000000`                 |
| `TIME '03:30'`                     | `12600000`            | `12600000`                | `12600000`                     | `12600000`                |
| `TIME '02:30 America/Los_Angeles'` | `37800000`            | `37800000`                | `37800000`                     | `37800000`                |
| `TIME '03:30 America/Los_Angeles'` | `41400000`            | `41400000`                | `41400000`                     | `41400000`                |

As you can see, there is no difference in representation of TIME WITH TIME ZONE, the difference is in the interpretation (casts, at_time_zone etc).
As mentioned in doc, TZ offset is lazy evaluated when required and the way it's evaluated changed.
**Once again I encourage you to read the doc, I'm quite sure that the idea was explained there, and my answers in PR are not as precise, as I spend much more time to refine this doc. I kinda lost parts of the context and I'm making mistakes explaining that to you here and trying to recall my intentions.**

 3. I don’t think it’s good idea. This would mean that some queries will fail depending on the data, which I think is something that should be avoided.

haozhun commented:
@fiedukow I'm very grateful for all the answers you provided. Thank you for taking the liberty to provide 32 (and not 16) answers for my second question. Many of the answers you provided there are different from what I thought. (The answer you provided are more desirable.) I'll read through the docs again.

fiedukow commented:
Any progress?

haozhun commented:
This PR is now on the top of my queue.

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java:
```diff
@@ -98,7 +98,7 @@ public static long currentTime(ConnectorSession session)
         // Stack value is number of milliseconds from start of the current day,
         // but the start of the day is relative to the current time zone.
-        long millis = getChronology(session.getTimeZoneKey()).millisOfDay().get(session.getStartTime());
+        long millis = UTC_CHRONOLOGY.millisOfDay().get(session.getStartTime());
```
This line has always raised concern and caused significant confusion.

Now, I finally realize that `current_time` has simply been completely wrong. And I realize that you also pointed this out in the document.

While I understand that Presto chose a very strange interpretation of time w/o tz, and timestamp w/o tz. It never occurred to me that existing `current_time` is simply completely wrong.

In addition, I also assumed the comment above is correct.

This fix deserves a lot more emphasis that it got. I hope you now understand why I'm asking all the questions and raising all the problems. It all comes from confusion as a result of reading this line.

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java:
```diff
@@ -98,7 +98,7 @@ public static long currentTime(ConnectorSession session)
     {
         // Stack value is number of milliseconds from start of the current day,
         // but the start of the day is relative to the current time zone.
```
Remove this incorrect comment

haozhun commented on presto-main/src/test/java/com/facebook/presto/operator/scalar/TestDateTimeFunctions.java:
```diff
@@ -143,15 +143,15 @@ private static long epochDaysInZone(TimeZoneKey timeZoneKey, long instant)
             throws Exception
     {
-        long millis = new LocalTime(session.getStartTime(), DATE_TIME_ZONE).getMillisOfDay();
+        long millis = new LocalTime(session.getStartTime(), DateTimeZone.UTC).getMillisOfDay();
```
I wonder if there is a better way to write this test. It doesn't need to be in this commit or this PR.

haozhun commented on presto-spi/src/main/java/com/facebook/presto/spi/ConnectorSession.java:
```diff
@@ -35,5 +35,8 @@ default String getUser()
     long getStartTime();
 
+    @Deprecated
+    boolean isLegacyTimestamp();
```
@martint Are you ok with this being in ConnectorSession. I feel like it's necessary and we don't have a choice.

haozhun commented on presto-spi/src/main/java/com/facebook/presto/spi/ConnectorSession.java:
```diff
@@ -35,5 +35,8 @@ default String getUser()
 
     long getStartTime();
 
+    @Deprecated
```
Don't mark as `@Deprecated`

haozhun commented on presto-main/src/main/java/com/facebook/presto/util/DateTimeUtils.java:
```diff
@@ -70,6 +70,8 @@ public static String printDate(int days)
         return DATE_FORMATTER.print(TimeUnit.DAYS.toMillis(days));
     }
 
+    @Deprecated
```
Don't mark as deprecated. This applies to all `@Deprecated` in this pull request.

I kind of see why you have `@Deprecated` after reading this rest of this pull request. You are using that to mark a method as legacy. You avoided renaming by using method overload for the new methods. To compensate for the confusion this could lead to, you then use `@Deprecated` to mark the legacy one. People reading the code at the callsites will have to rely on strike-through lines in the IDE to distinguish between the two.

I see that all the deprecated ones takes a tz, while new ones don't. This is a little bit subtle in my taste.

An alternative is to rename the old method through with two steps:
* commit 1 renames `xyz` to `legacyXyz`.
* commit 2 changes `legacyXyz` to use `LEGACY_TIMESTAMP_WITHOUT_TIME_ZONE_FORMATTER`, and adds back `xyz` with new semantics.

I prefer renaming. What do you say @martint?

haozhun commented on presto-spi/src/main/java/com/facebook/presto/spi/type/SqlTime.java:
```diff
@@ -17,32 +17,48 @@
+        this.sessionTimeZoneKey = Optional.empty();
+    }
+
+    @Deprecated
```
Instead of having two overloads of `SqlTime` (with one of them marked `@Deprecated`, what about having two factories: `SqlTime.of` and `SqlTime.ofLegacy`?

(Again, I prefer that. But this is not something I absolutely can't tolerate. I'll need to see what other people think.)

haozhun commented on presto-spi/src/main/java/com/facebook/presto/spi/type/SqlTime.java:
```diff
@@ -17,32 +17,48 @@
 
-    public TimeZoneKey getSessionTimeZoneKey()
+    @Deprecated
+    public Optional<TimeZoneKey> getSessionTimeZoneKey()
```
Depending on how you are using this. It may (and may not) be reasonable to do this instead:

```
public TimeZoneKey getSessionTimeZoneKey() {
  checkState(sessionTimeZoneKey.isPresent());
  return sessionTimeZoneKey.get();
}

haozhun commented on presto-spi/src/main/java/com/facebook/presto/spi/type/SqlTimestamp.java:
```diff
@@ -67,6 +84,12 @@ public boolean equals(Object obj)
+            return Instant.ofEpochMilli(millisUtc).atZone(ZoneId.of(sessionTimeZoneKey.get().getId())).format(formatter);
+        }
+        else {
+            // FIXME this should be exact equivalent of TIMESTAMP to VARCHAR cast
```
Resolve this `FIXME`?

haozhun commented on presto-spi/src/main/java/com/facebook/presto/spi/type/SqlTimestamp.java:
```diff
@@ -19,30 +19,47 @@
+        sessionTimeZoneKey = Optional.empty();
+    }
 
+    @Deprecated
```
Comments for `SqlTime` also applies to this class.

haozhun commented:
Everything looks pretty reasonable so far. I'll continue on later. Please let me know about your opinion on my comments so far.

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java:
```diff
@@ -96,9 +98,17 @@ public static long currentDate(ConnectorSession session)
+        // and we need to have UTC millis for packDateTimeWithZone
         long millis = UTC_CHRONOLOGY.millisOfDay().get(session.getStartTime());
+
+        if (!session.isLegacyTimestamp()) {
```
I'll defer review related to `time with time zone` to a later time.

haozhun commented on presto-main/src/test/java/com/facebook/presto/operator/scalar/TestDateTimeFunctions.java:
```diff
@@ -143,29 +149,102 @@ private static long epochDaysInZone(TimeZoneKey timeZoneKey, long instant)
+    public void testCurrentTimeLegacy()
+            throws Exception
+    {
+        Session localSession = testSessionBuilder()
```
I'll defer review related to `time with time zone` to a later time.

haozhun commented on presto-main/src/test/java/com/facebook/presto/operator/scalar/TestDateTimeFunctions.java:
```diff
@@ -143,29 +149,102 @@ private static long epochDaysInZone(TimeZoneKey timeZoneKey, long instant)
     {
-        long millis = new LocalTime(session.getStartTime(), DateTimeZone.UTC).getMillisOfDay();
-        functionAssertions.assertFunction("CURRENT_TIME", TIME_WITH_TIME_ZONE, new SqlTimeWithTimeZone(millis, session.getTimeZoneKey()));
+        Session localSession = testSessionBuilder()
```
I'll defer review related to `time with time zone` to a later time.

haozhun commented on presto-main/src/test/java/com/facebook/presto/operator/scalar/TestDateTimeFunctions.java:
```diff
@@ -143,29 +149,102 @@ private static long epochDaysInZone(TimeZoneKey timeZoneKey, long instant)
     {
-        long millis = new LocalTime(session.getStartTime(), DateTimeZone.UTC).getMillisOfDay();
-        functionAssertions.assertFunction("LOCALTIME", TimeType.TIME, toTime(millis));
+        Session localSession = testSessionBuilder()
```
Add a test (in maybe a separate test method) for 2017-03-12 1:30 and 3:30 in America/Los_Angeles

haozhun commented on presto-main/src/main/java/com/facebook/presto/testing/TestingSqlTime.java:
```diff
@@ -0,0 +1,73 @@
+    private TestingSqlTime()
+    {}
+
+    public static SqlTimestamp sqlTimestampOf(int year,
```
first argument on its own line

haozhun commented:
"Change default TIME/TIMESTAMP semantics to ANSI SQL" - looks good

haozhun commented on presto-main/src/test/java/com/facebook/presto/type/TestDateTimeOperators.java:
```diff
@@ -71,4 +84,89 @@ public void testDaylightTimeSavingSwitchCrossingIsNotApplied()
+    public void testTimeWithTimeZoneRepresentation()
+    {
+        // PST -> PDT date
+        testTimeRepresentationOnDate(new DateTime(2017, 3, 12, 10, 0, 0, 0, DateTimeZone.UTC),
```
First argument on its own line. This comments applies to multiple places in this commit.

haozhun commented:
For "time with time zone", I believe we still have disagreements.

* Commit "Fix at_timezone(TIME WITH TIME ZONE)" is about that type.
* "Fix current_time, localtime & localtimestamp semantics" involves that.

For the others, I have some comments. But they look mostly good. I propose we get those comments addressed so that we can have those merged.

For "time with time zone", here are my opinions:

* The old behavior is completely broken. And maintaining compatibility is super complicated in this case, I propose just fixing it.
* The "utc-unixtime + tz" representation inherently depends on an implicit date (whether it's 1970-01-01 or today). I propose a "local-unixtime + tz" representation. This way, code will be easier to read and reason about.
* For at_timezone and comparisons, I think we should allow all operations on offset zones. And if source/target (in the case of at_timezone) or left/right (in the case of comparison) has different political zones, let's just fail it. Reasoning about how it work will be messy. And there is no good answer regardless of how hard we try.

fiedukow commented:
Thanks for the first pass of the review.
I'm currently working on other TS related problem caused by incompatibility of TIMESTAMP semantics after this patch with TIMESTAMP semantics in Hive.
I'll rebase this PR soon, but will have to add some more testing and additional fix to actually make this mergeable.
I'll get to fixing your comments just after adding this patch.

haozhun commented:
Please assign back to me when this is ready for review again.

fiedukow commented on presto-spi/src/main/java/com/facebook/presto/spi/type/TimestampType.java:
```diff
@@ -39,7 +39,12 @@ public Object getObjectValue(ConnectorSession session, Block block, int position
         }
 
-        return new SqlTimestamp(block.getLong(position, 0), session.getTimeZoneKey());
+        if (session.isLegacyTimestamp()) {
```
Comment in header of this class is invalid and should be changed.

losipiuk commented:
Superseded by https://github.com/prestodb/presto/pull/9385
